### PR TITLE
Adding Unavailable Menu Item docs

### PR DESCRIPTION
**Pull request by ktabors**
## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [X] Filled out test instructions.
- [X] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Goto the MenuTrigger Docs
Review the "Unavailable Menu Item" section

https://reactspectrum.blob.core.windows.net/reactspectrum/bda56c31b7f5e19eb72b96f8592e460d8c985b4b/docs/react-spectrum/MenuTrigger.html#unavailable-items

## 🧢 Your Project:
RSP

### CODE REVIEW

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx [hunk 1]
```diff
@@ -121,6 +121,63 @@ import {Text} from '@react-spectrum/text';
 </MenuTrigger>
 ```
 
+## Unavilalbe Menu Item <VersionBadge version="alpha" style={{marginLeft: 4, verticalAlign: 'bottom'}} />
```
Comment (majornista) [inline packages/@react-spectrum/menu/docs/MenuTrigger.mdx:124]:
```suggestion
## Unavailable Menu Item <VersionBadge version="alpha" style={{marginLeft: 4, verticalAlign: 'bottom'}} />
```

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx [hunk 2]
```diff
@@ -121,6 +121,63 @@ import {Text} from '@react-spectrum/text';
...
+
+ContextualHelpTrigger shows a user extra information about the state of an unavilalbe Menu Item.
```
Comment (majornista) [inline packages/@react-spectrum/menu/docs/MenuTrigger.mdx:126]:
```suggestion
ContextualHelpTrigger shows a user extra information about the state of an unavailable Menu Item.
```

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx [hunk 3]
```diff
@@ -121,6 +121,63 @@ import {Text} from '@react-spectrum/text';
...
+It provides context for the state and may point users to more information elsewhere.
+
+The ContextualHelpTrigger accepts exactly
+two children: the Item which triggers opening of the Dialog and the Dialog itself. The trigger must be
+the first child passed into the ContextualHelpTrigger and should be an Item. Similar to
+ContextualHelp, the layout of the Dialog is very diliberate. See [ContextualHelp](ContextualHelp.html#content) for further explanation.
```
Comment (majornista) [inline packages/@react-spectrum/menu/docs/MenuTrigger.mdx:132]:
```suggestion
ContextualHelp, the layout of the Dialog is very deliberate. See [ContextualHelp](ContextualHelp.html#content) for further explanation.
```

**Review by majornista — DISMISSED** [review]
Should this be under Menu in addition to or rather than MenuTrigger?

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx (hunk 2 above)
Comment (reidbarber) [inline packages/@react-spectrum/menu/docs/MenuTrigger.mdx:126]: I would probably include something about why you would want to use this (when you want to keep an item in the UI, but it isn't available to a user + you have the ability to add context as to why it's unavailable). This currently sounds like there is another way to make items unavailable, but this is how you add extra information.

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx [hunk 4]
```diff
@@ -121,6 +121,63 @@ import {Text} from '@react-spectrum/text';
...
+
+By default, a ContextualHelpTrigger's Dialog is opened by hovering or pressing the trigger element or activating
+it via the <Keyboard>Space</Keyboard>, <Keyboard>Enter</Keyboard>, or <Keyboard>Right Arrow</Keyboard> keys.
+Hovering another item or pressing the <Keyboard>Esc</Keyboard> key will close the Dialog and leave the Menu open.
+
+When Menu Items are unavailabe via a ContextualHelpTrigger, the Menu `onAction` and `onSelectionChange`
+callbacks will not occur.
+
+Setting the `isUnavailable` prop on the ContextualHelpTrigger disables a Menu Item.
```
Comment (reidbarber) [inline packages/@react-spectrum/menu/docs/MenuTrigger.mdx:141]: I would probably include more info on this use case (ability to have this be dynamic).

**Review by reidbarber — COMMENTED** [review]
I'm not seeing the popover in the docs for some reason.

**Comment by ktabors** [comment]
> Should this be under Menu in addition to or rather than MenuTrigger?

Rob suggested this be added to MenuTrigger and we later Merge Menu and MenuTrigger docs.

**Comment by ktabors** [comment]
> I'm not seeing the popover in the docs for some reason.

This is the bug Rob warned us about. The position is wrong.

**Review by reidbarber — APPROVED** [review]

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx [hunk 5]
```diff
@@ -121,6 +121,63 @@ import {Text} from '@react-spectrum/text';
...
+two children: the Item which triggers opening of the Dialog and the Dialog itself. The trigger must be
+the first child passed into the ContextualHelpTrigger and should be an Item. Similar to
+ContextualHelp, the layout of the Dialog is very deliberate. See [ContextualHelp](ContextualHelp.html#content) for further explanation.
+
+By default, a ContextualHelpTrigger's Dialog is opened by hovering, pressing the trigger element or activating
+it via the <Keyboard>Space</Keyboard>, <Keyboard>Enter</Keyboard>, or <Keyboard>Right Arrow</Keyboard> keys.
+Hovering another item or pressing the <Keyboard>Esc</Keyboard> key will close the Dialog and leave the Menu open.
+
+When Menu Items are unavailable via a ContextualHelpTrigger, the Menu `onAction` and `onSelectionChange`
+callbacks will not occur.
```
Comment (LFDanLu) [inline packages/@react-spectrum/menu/docs/MenuTrigger.mdx:139]: nit: maybe something like this? It previously sounded like onAction/onSelectionChange wouldn't fire for any of the items even if 
```suggestion
Note that the Menu's `onAction` and `onSelectionChange`
callbacks will not fire for items made unavailable by a ContextualHelpTrigger.
```

Comment (LFDanLu) [reply packages/@react-spectrum/menu/docs/MenuTrigger.mdx:139]: Or maybe move this sentence after the one below (aka the "Setting the `isUnavailable` prop on the...")?

**Review by LFDanLu — COMMENTED** [review]
Just one comment, but otherwise LGTM

**Review by reidbarber — APPROVED** [review]

**Review by LFDanLu — CHANGES_REQUESTED** [review]
Positioning looks good to me now, but looking at the docs again raised some questions for me:
- Should ContextualHelpTrigger's `props` actually include `ItemProps`? Feels like it should only expose `isUnavailable` since the other props included by `ItemProps` don't actually do anything (they don't get propagated to the wrapped Item I believe)
- Probably should include ContextualHelpTrigger in the  `componentNames={['MenuTrigger', 'Menu']}` and add a PropTable for it?

What do you think?

File: packages/@react-spectrum/menu/docs/types.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import {ReactElement} from 'react';
+
+// The doc generator is not smart enough to handle the real types for ContextualHelpTrigger so this is a simpler one.
+export interface ContextualHelpTriggerProps {
+  /** Whether the menu item is currently unavailable. */
+  isUnavailable?: boolean,
+  /** The Item which triggers opening of the Dialog, and the Dialog itself. */
+  children: [ReactElement, ReactElement]
+}
```
Comment (LFDanLu) [inline packages/@react-spectrum/menu/docs/types.ts:9]: Sorry about the back and forth with this, but perhaps we should move this interface definition into ContextualHelpTrigger and set it as the SpectrumMenuDialogTriggerProps so the exported props match the actual props the user would provide? Also we'd be able to use it directly as the `let _Item = ContextualHelpTrigger...` props and thus wouldn't risk drift between this doc definition here and the `Pick` set in ContextualHelpTrigger?

aka:

```
// In packages/@react-spectrum/menu/src/ContextualHelpTrigger.tsx
interface MenuDialogTriggerProps<T> extends ItemProps<T> {
  isUnavailable?: boolean,
  targetKey: Key
}

export interface SpectrumMenuDialogTriggerProps extends Pick<MenuDialogTriggerProps<T>, 'isUnavailable' | 'children'>) {}

function ContextualHelpTrigger<T>(props: MenuDialogTriggerProps<T>): ReactElement {
...

let _Item = ContextualHelpTrigger as <T>(props: SpectrumMenuDialogTriggerProps<T>) => JSX.Element;
```
and the mdx file could do something like:

```
import contextualHelpTriggerTypes from 'docs:@react-spectrum/menu/src/ContextualHelpTrigger.tsx';

// Similar to how we import Column props in TableView.mdx
<PropTable component={{
  props: {
    properties: Object.fromEntries(Object.entries(menu.exports.SpectrumMenuDialogTriggerProps.properties))
  }
}} links={contextualHelpTriggerTypes.links} style={{marginBottom: '40px'}} />

```

I forget if there is an easier way to grab the types here, maybe `docs.exports.SpectrumMenuDialogTriggerProps` works?

Comment (reidbarber) [reply packages/@react-spectrum/menu/docs/types.ts]: I think I managed to get it to work. When using Pick I wasn't seeing anything in the props table, so I updated to use Omit instead and rearranged some things.

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx [hunk 7]
```diff
@@ -121,10 +122,75 @@ import {Text} from '@react-spectrum/text';
 </MenuTrigger>
 ```
 
+## Unavailable Items <VersionBadge version="alpha" style={{marginLeft: 4, verticalAlign: 'bottom'}} />
+
+ContextualHelpTrigger allows making a Menu Item unavailable along with showing a user extra information about the state of that unavailable item.
+It provides context and may point users to more information elsewhere.
```
Comment (snowystinger) [inline packages/@react-spectrum/menu/docs/MenuTrigger.mdx:128]:
```suggestion
ContextualHelpTrigger disables a menu item's action and replaces it with a popover with information on why the item is unavailable and may link users to more information elsewhere.
```

Would this be better?

File: packages/@react-spectrum/menu/src/ContextualHelpTrigger.tsx [hunk 8]
```diff
@@ -19,12 +19,17 @@ import {Modal, Popover} from '@react-spectrum/overlays';
 import React, {Key, ReactElement, useRef} from 'react';
 import {useOverlayTriggerState} from '@react-stately/overlays';
 
-export interface SpectrumMenuDialogTriggerProps<T> extends ItemProps<T> {
+interface MenuDialogTriggerProps<T> extends ItemProps<T> {
+  /** Whether the menu item is currently unavailable. */
   isUnavailable?: boolean,
+  /** The Item which triggers opening of the Dialog, and the Dialog itself. */
```
Comment (snowystinger) [inline packages/@react-spectrum/menu/src/ContextualHelpTrigger.tsx:25]:
```suggestion
  /** The triggering Item and the Dialog, respectively. */
```

File: packages/@react-spectrum/menu/docs/MenuTrigger.mdx (hunk 7 above)
Comment (reidbarber) [reply packages/@react-spectrum/menu/docs/MenuTrigger.mdx]: Updated

**Review by snowystinger — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]

**Pull request merged**